// vscode/dist/configuration.js
import { J as IUserDataProfilesService, n as IFileService, E as Emitter, _ as __decorate, a as __param, o as IWorkbenchEnvironmentService } from './services2.js';
import { o as onServicesInitialized, P as Position, I as IPolicyService, k as NullPolicyService, l as NullPolicyConfiguration, q as PolicyConfiguration, U as UserSettings } from './polyfill.js';
import { bp as IExtensionService, bL as ILifecycleService } from './missing-services.js';
import { URI } from 'monaco-editor/esm/vs/base/common/uri.js';
import { Disposable } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { DefaultConfiguration } from 'monaco-editor/esm/vs/platform/configuration/common/configurations.js';
import { extUriBiasedIgnorePathCase } from 'monaco-editor/esm/vs/base/common/resources.js';
import { Configuration, ConfigurationModel, ConfigurationChangeEvent } from 'monaco-editor/esm/vs/platform/configuration/common/configurationModels.js';
import { RunOnceScheduler, Delayer } from 'monaco-editor/esm/vs/base/common/async.js';
import { IConfigurationService } from 'monaco-editor/esm/vs/platform/configuration/common/configuration.js';
import { IModelService } from 'monaco-editor/esm/vs/editor/common/services/model.js';
import { ILanguageService } from 'monaco-editor/esm/vs/editor/common/languages/language.js';
import { d as defaultSettingsSchemaId, u as userSettingsSchemaId, p as profileSettingsSchemaId, m as machineSettingsSchemaId, w as workspaceSettingsSchemaId, f as folderSettingsSchemaId, s as setProperty } from './configuration2.js';
import { k as keyFromOverrideIdentifiers, c as configurationDefaultsSchemaId } from './configurationRegistry.js';
export { C as ConfigurationScope } from './configurationRegistry.js';
import { Event } from 'monaco-editor/esm/vs/base/common/event.js';
import { Extensions, applicationSettings, windowSettings, resourceSettings, allSettings, machineSettings, machineOverridableSettings, OVERRIDE_PROPERTY_PATTERN, resourceLanguageSettingsSchemaId } from 'monaco-editor/esm/vs/platform/configuration/common/configurationRegistry.js';
import { localize } from 'monaco-editor/esm/vs/nls.js';
import { IWorkspaceContextService } from 'monaco-editor/esm/vs/platform/workspace/common/workspace.js';
import { IWorkspaceTrustManagementService } from 'monaco-editor/esm/vs/platform/workspace/common/workspaceTrust.js';
import { Registry } from 'monaco-editor/esm/vs/platform/registry/common/platform.js';
import { Extensions as Extensions$1 } from 'monaco-editor/esm/vs/platform/jsonschemas/common/jsonContributionRegistry.js';
import { g as getServiceOverride$2 } from './workspaceContext.js';
import { g as getServiceOverride$1 } from './files.js';
import { SyncDescriptor } from 'monaco-editor/esm/vs/platform/instantiation/common/descriptors.js';
import { ITextResourceConfigurationService } from 'monaco-editor/esm/vs/editor/common/services/textResourceConfiguration.js';
import { StandaloneServices } from 'monaco-editor/esm/vs/editor/standalone/browser/standaloneServices.js';
import { VSBuffer } from 'monaco-editor/esm/vs/base/common/buffer.js';
import { createTextBuffer, TextModel } from 'monaco-editor/esm/vs/editor/common/model/textModel.js';
import { ILogService } from 'monaco-editor/esm/vs/platform/log/common/log.js';
function isConfigurationOverrides(thing) {
    return thing
        && typeof thing === 'object'
        && (!thing.overrideIdentifier || typeof thing.overrideIdentifier === 'string')
        && (!thing.resource || thing.resource instanceof URI);
}
function isConfigurationUpdateOverrides(thing) {
    return thing
        && typeof thing === 'object'
        && (!thing.overrideIdentifiers || Array.isArray(thing.overrideIdentifiers))
        && !thing.overrideIdentifier
        && (!thing.resource || thing.resource instanceof URI);
}
class ConfigurationService extends Disposable {
    constructor(settingsResource, fileService, policyService, logService) {
        super();
        this.settingsResource = settingsResource;
        this._onDidChangeConfiguration = this._register(( (new Emitter())));
        this.onDidChangeConfiguration = this._onDidChangeConfiguration.event;
        this.defaultConfiguration = this._register(( (new DefaultConfiguration())));
        this.policyConfiguration = policyService instanceof NullPolicyService ? ( (new NullPolicyConfiguration())) : this._register(( (new PolicyConfiguration(this.defaultConfiguration, policyService, logService))));
        this.userConfiguration = this._register(( (new UserSettings(this.settingsResource, undefined, extUriBiasedIgnorePathCase, fileService))));
        this.configuration = ( (new Configuration(
            this.defaultConfiguration.configurationModel,
            this.policyConfiguration.configurationModel,
             (new ConfigurationModel()),
             (new ConfigurationModel())
        )));
        this.reloadConfigurationScheduler = this._register(( (new RunOnceScheduler(() => this.reloadConfiguration(), 50))));
        this._register(this.defaultConfiguration.onDidChangeConfiguration(({ defaults, properties }) => this.onDidDefaultConfigurationChange(defaults, properties)));
        this._register(this.policyConfiguration.onDidChangeConfiguration(model => this.onDidPolicyConfigurationChange(model)));
        this._register(this.userConfiguration.onDidChange(() => this.reloadConfigurationScheduler.schedule()));
    }
    async initialize() {
        const [defaultModel, policyModel, userModel] = await Promise.all([this.defaultConfiguration.initialize(), this.policyConfiguration.initialize(), this.userConfiguration.loadConfiguration()]);
        this.configuration = ( (new Configuration(
            defaultModel,
            policyModel,
             (new ConfigurationModel()),
            userModel
        )));
    }
    getConfigurationData() {
        return this.configuration.toData();
    }
    getValue(arg1, arg2) {
        const section = typeof arg1 === 'string' ? arg1 : undefined;
        const overrides = isConfigurationOverrides(arg1) ? arg1 : isConfigurationOverrides(arg2) ? arg2 : {};
        return this.configuration.getValue(section, overrides, undefined);
    }
    updateValue(key, value, arg3, arg4) {
        return Promise.reject(( (new Error('not supported'))));
    }
    inspect(key) {
        return this.configuration.inspect(key, {}, undefined);
    }
    keys() {
        return (
             (this.configuration.keys(undefined))
        );
    }
    async reloadConfiguration() {
        const configurationModel = await this.userConfiguration.loadConfiguration();
        this.onDidChangeUserConfiguration(configurationModel);
    }
    onDidChangeUserConfiguration(userConfigurationModel) {
        const previous = this.configuration.toData();
        const change = this.configuration.compareAndUpdateLocalUserConfiguration(userConfigurationModel);
        this.trigger(change, previous, 2 );
    }
    onDidDefaultConfigurationChange(defaultConfigurationModel, properties) {
        const previous = this.configuration.toData();
        const change = this.configuration.compareAndUpdateDefaultConfiguration(defaultConfigurationModel, properties);
        this.trigger(change, previous, 7 );
    }
    onDidPolicyConfigurationChange(policyConfiguration) {
        const previous = this.configuration.toData();
        const change = this.configuration.compareAndUpdatePolicyConfiguration(policyConfiguration);
        this.trigger(change, previous, 7 );
    }
    trigger(configurationChange, previous, source) {
        const event = ( (new ConfigurationChangeEvent(configurationChange, { data: previous }, this.configuration)));
        event.source = source;
        event.sourceConfig = this.getTargetConfiguration(source);
        this._onDidChangeConfiguration.fire(event);
    }
    getTargetConfiguration(target) {
        switch (target) {
            case 7 :
                return this.configuration.defaults.contents;
            case 2 :
                return this.configuration.localUserConfiguration.contents;
        }
        return {};
    }
}
let TextResourceConfigurationService = class TextResourceConfigurationService extends Disposable {
    constructor(configurationService, modelService, languageService) {
        super();
        this.configurationService = configurationService;
        this.modelService = modelService;
        this.languageService = languageService;
        this._onDidChangeConfiguration = this._register(( (new Emitter())));
        this.onDidChangeConfiguration = this._onDidChangeConfiguration.event;
        this._register(this.configurationService.onDidChangeConfiguration(e => this._onDidChangeConfiguration.fire(this.toResourceConfigurationChangeEvent(e))));
    }
    getValue(resource, arg2, arg3) {
        if (typeof arg3 === 'string') {
            return this._getValue(resource, Position.isIPosition(arg2) ? arg2 : null, arg3);
        }
        return this._getValue(resource, null, typeof arg2 === 'string' ? arg2 : undefined);
    }
    updateValue(resource, key, value, configurationTarget) {
        const language = this.getLanguage(resource, null);
        const configurationValue = this.configurationService.inspect(key, { resource, overrideIdentifier: language });
        if (configurationTarget === undefined) {
            configurationTarget = this.deriveConfigurationTarget(configurationValue, language);
        }
        switch (configurationTarget) {
            case 8 :
                return this._updateValue(key, value, configurationTarget, configurationValue.memory?.override, resource, language);
            case 6 :
                return this._updateValue(key, value, configurationTarget, configurationValue.workspaceFolder?.override, resource, language);
            case 5 :
                return this._updateValue(key, value, configurationTarget, configurationValue.workspace?.override, resource, language);
            case 4 :
                return this._updateValue(key, value, configurationTarget, configurationValue.userRemote?.override, resource, language);
            default:
                return this._updateValue(key, value, configurationTarget, configurationValue.userLocal?.override, resource, language);
        }
    }
    _updateValue(key, value, configurationTarget, overriddenValue, resource, language) {
        if (language && overriddenValue !== undefined) {
            return this.configurationService.updateValue(key, value, { resource, overrideIdentifier: language }, configurationTarget);
        }
        else {
            return this.configurationService.updateValue(key, value, { resource }, configurationTarget);
        }
    }
    deriveConfigurationTarget(configurationValue, language) {
        if (language) {
            if (configurationValue.memory?.override !== undefined) {
                return 8 ;
            }
            if (configurationValue.workspaceFolder?.override !== undefined) {
                return 6 ;
            }
            if (configurationValue.workspace?.override !== undefined) {
                return 5 ;
            }
            if (configurationValue.userRemote?.override !== undefined) {
                return 4 ;
            }
            if (configurationValue.userLocal?.override !== undefined) {
                return 3 ;
            }
        }
        if (configurationValue.memory?.value !== undefined) {
            return 8 ;
        }
        if (configurationValue.workspaceFolder?.value !== undefined) {
            return 6 ;
        }
        if (configurationValue.workspace?.value !== undefined) {
            return 5 ;
        }
        if (configurationValue.userRemote?.value !== undefined) {
            return 4 ;
        }
        return 3 ;
    }
    _getValue(resource, position, section) {
        const language = resource ? this.getLanguage(resource, position) : undefined;
        if (typeof section === 'undefined') {
            return this.configurationService.getValue({ resource, overrideIdentifier: language });
        }
        return this.configurationService.getValue(section, { resource, overrideIdentifier: language });
    }
    getLanguage(resource, position) {
        const model = this.modelService.getModel(resource);
        if (model) {
            return position ? model.getLanguageIdAtPosition(position.lineNumber, position.column) : model.getLanguageId();
        }
        return this.languageService.guessLanguageIdByFilepathOrFirstLine(resource);
    }
    toResourceConfigurationChangeEvent(configurationChangeEvent) {
        return {
            affectedKeys: configurationChangeEvent.affectedKeys,
            affectsConfiguration: (resource, configuration) => {
                const overrideIdentifier = resource ? this.getLanguage(resource, null) : undefined;
                return configurationChangeEvent.affectsConfiguration(configuration, { resource, overrideIdentifier });
            }
        };
    }
};
TextResourceConfigurationService = ( (__decorate([
    ( (__param(0, IConfigurationService))),
    ( (__param(1, IModelService))),
    ( (__param(2, ILanguageService)))
], TextResourceConfigurationService)));
let RegisterConfigurationSchemasContribution = class RegisterConfigurationSchemasContribution extends Disposable {
    constructor(workspaceContextService, environmentService, workspaceTrustManagementService, extensionService, lifecycleService) {
        super();
        this.workspaceContextService = workspaceContextService;
        this.environmentService = environmentService;
        this.workspaceTrustManagementService = workspaceTrustManagementService;
        extensionService.whenInstalledExtensionsRegistered().then(() => {
            this.registerConfigurationSchemas();
            const configurationRegistry = ( (Registry.as(Extensions.Configuration)));
            const delayer = this._register(( (new Delayer(50))));
            this._register(Event.any(configurationRegistry.onDidUpdateConfiguration, configurationRegistry.onDidSchemaChange, workspaceTrustManagementService.onDidChangeTrust)(() => delayer.trigger(() => this.registerConfigurationSchemas(), lifecycleService.phase === 4  ? undefined : 2500 )));
        });
    }
    registerConfigurationSchemas() {
        const jsonRegistry = ( (Registry.as(Extensions$1.JSONContribution)));
        const allSettingsSchema = {
            properties: allSettings.properties,
            patternProperties: allSettings.patternProperties,
            additionalProperties: true,
            allowTrailingCommas: true,
            allowComments: true
        };
        const userSettingsSchema = this.environmentService.remoteAuthority ?
            {
                properties: Object.assign({}, applicationSettings.properties, windowSettings.properties, resourceSettings.properties),
                patternProperties: allSettings.patternProperties,
                additionalProperties: true,
                allowTrailingCommas: true,
                allowComments: true
            }
            : allSettingsSchema;
        const profileSettingsSchema = {
            properties: Object.assign({}, machineSettings.properties, machineOverridableSettings.properties, windowSettings.properties, resourceSettings.properties),
            patternProperties: allSettings.patternProperties,
            additionalProperties: true,
            allowTrailingCommas: true,
            allowComments: true
        };
        const machineSettingsSchema = {
            properties: Object.assign({}, machineSettings.properties, machineOverridableSettings.properties, windowSettings.properties, resourceSettings.properties),
            patternProperties: allSettings.patternProperties,
            additionalProperties: true,
            allowTrailingCommas: true,
            allowComments: true
        };
        const workspaceSettingsSchema = {
            properties: Object.assign({}, this.checkAndFilterPropertiesRequiringTrust(machineOverridableSettings.properties), this.checkAndFilterPropertiesRequiringTrust(windowSettings.properties), this.checkAndFilterPropertiesRequiringTrust(resourceSettings.properties)),
            patternProperties: allSettings.patternProperties,
            additionalProperties: true,
            allowTrailingCommas: true,
            allowComments: true
        };
        jsonRegistry.registerSchema(defaultSettingsSchemaId, {
            properties: ( (Object.keys(allSettings.properties))).reduce((result, key) => {
                result[key] = Object.assign({ deprecationMessage: undefined }, allSettings.properties[key]);
                return result;
            }, {}),
            patternProperties: ( (Object.keys(allSettings.patternProperties))).reduce((result, key) => {
                result[key] = Object.assign({ deprecationMessage: undefined }, allSettings.patternProperties[key]);
                return result;
            }, {}),
            additionalProperties: true,
            allowTrailingCommas: true,
            allowComments: true
        });
        jsonRegistry.registerSchema(userSettingsSchemaId, userSettingsSchema);
        jsonRegistry.registerSchema(profileSettingsSchemaId, profileSettingsSchema);
        jsonRegistry.registerSchema(machineSettingsSchemaId, machineSettingsSchema);
        if (3  === this.workspaceContextService.getWorkbenchState()) {
            const folderSettingsSchema = {
                properties: Object.assign({}, this.checkAndFilterPropertiesRequiringTrust(machineOverridableSettings.properties), this.checkAndFilterPropertiesRequiringTrust(resourceSettings.properties)),
                patternProperties: allSettings.patternProperties,
                additionalProperties: true,
                allowTrailingCommas: true,
                allowComments: true
            };
            jsonRegistry.registerSchema(workspaceSettingsSchemaId, workspaceSettingsSchema);
            jsonRegistry.registerSchema(folderSettingsSchemaId, folderSettingsSchema);
        }
        else {
            jsonRegistry.registerSchema(workspaceSettingsSchemaId, workspaceSettingsSchema);
            jsonRegistry.registerSchema(folderSettingsSchemaId, workspaceSettingsSchema);
        }
        jsonRegistry.registerSchema(configurationDefaultsSchemaId, {
            type: 'object',
            description: ( (localize(
                'configurationDefaults.description',
                'Contribute defaults for configurations'
            ))),
            properties: Object.assign({}, machineOverridableSettings.properties, windowSettings.properties, resourceSettings.properties),
            patternProperties: {
                [OVERRIDE_PROPERTY_PATTERN]: {
                    type: 'object',
                    default: {},
                    $ref: resourceLanguageSettingsSchemaId,
                }
            },
            additionalProperties: false
        });
    }
    checkAndFilterPropertiesRequiringTrust(properties) {
        if (this.workspaceTrustManagementService.isWorkspaceTrusted()) {
            return properties;
        }
        const result = {};
        Object.entries(properties).forEach(([key, value]) => {
            if (!value.restricted) {
                result[key] = value;
            }
        });
        return result;
    }
};
RegisterConfigurationSchemasContribution = ( (__decorate([
    ( (__param(0, IWorkspaceContextService))),
    ( (__param(1, IWorkbenchEnvironmentService))),
    ( (__param(2, IWorkspaceTrustManagementService))),
    ( (__param(3, IExtensionService))),
    ( (__param(4, ILifecycleService)))
], RegisterConfigurationSchemasContribution)));
async function updateUserConfiguration(configurationJson) {
    const userDataProfilesService = StandaloneServices.get(IUserDataProfilesService);
    await StandaloneServices.get(IFileService).writeFile(userDataProfilesService.defaultProfile.settingsResource, VSBuffer.fromString(configurationJson));
}
async function getUserConfiguration() {
    const userDataProfilesService = StandaloneServices.get(IUserDataProfilesService);
    return ( (await StandaloneServices.get(IFileService).readFile(userDataProfilesService.defaultProfile.settingsResource)).value.toString());
}
function onUserConfigurationChange(callback) {
    const userDataProfilesService = StandaloneServices.get(IUserDataProfilesService);
    return StandaloneServices.get(IFileService).onDidFilesChange(e => {
        if (e.affects(userDataProfilesService.defaultProfile.settingsResource)) {
            callback();
        }
    });
}
const configurationRegistry = ( Registry.as(Extensions.Configuration));
function updateJsonValue(jsonValue, jsonPath, value, configFileUrl) {
    const creationOptions = StandaloneServices.get(IModelService).getCreationOptions('json', configFileUrl, true);
    const { textBuffer, disposable: bufferDisposable } = createTextBuffer(jsonValue, creationOptions.defaultEOL);
    try {
        const options = TextModel.resolveOptions(textBuffer, creationOptions);
        const edits = setProperty(jsonValue, jsonPath, value, options);
        return edits.reduce((config, edit) => {
            return `${config.slice(0, edit.offset)}${edit.content}${config.slice(edit.offset + edit.length)}`;
        }, jsonValue);
    }
    finally {
        bufferDisposable.dispose();
    }
}
let InjectedConfigurationService = class InjectedConfigurationService extends ConfigurationService {
    constructor(userDataProfilesService, fileService, policyService, logService) {
        super(userDataProfilesService.defaultProfile.settingsResource, fileService, policyService, logService);
        this.userDataProfilesService = userDataProfilesService;
    }
    getValue(...args) {
        const value = super.getValue.apply(this, args);
        if (args[0] === 'keyboard') {
            return value ?? {};
        }
        return value;
    }
    async updateValue(key, value, arg3, arg4) {
        const overrides = isConfigurationUpdateOverrides(arg3)
            ? arg3
            : isConfigurationOverrides(arg3) ? { resource: arg3.resource, overrideIdentifiers: arg3.overrideIdentifier != null ? [arg3.overrideIdentifier] : undefined } : undefined;
        const target = (overrides != null ? arg4 : arg3) ?? 2 ;
        if (target === 8 ) {
            return this['configuration'].updateValue(key, value, overrides);
        }
        if (target === 2 ) {
            const userConfiguration = await getUserConfiguration();
            const jsonPath = overrides?.overrideIdentifiers != null && overrides.overrideIdentifiers.length > 0 ? [keyFromOverrideIdentifiers(overrides.overrideIdentifiers), key] : [key];
            const newConfiguration = updateJsonValue(userConfiguration, jsonPath, value, this.userDataProfilesService.defaultProfile.settingsResource);
            await updateUserConfiguration(newConfiguration);
            return;
        }
        return Promise.reject(new Error('not supported'));
    }
};
InjectedConfigurationService = ( __decorate([
    ( __param(0, IUserDataProfilesService)),
    ( __param(1, IFileService)),
    ( __param(2, IPolicyService)),
    ( __param(3, ILogService))
], InjectedConfigurationService));
async function initialize(instantiationService) {
    instantiationService.createInstance(RegisterConfigurationSchemasContribution);
    const configurationService = instantiationService.invokeFunction((accessor) => accessor.get(IConfigurationService));
    await Promise.all([
        updateUserConfiguration('{}'),
        configurationService.initialize()
    ]);
}
function getServiceOverride() {
    onServicesInitialized(initialize);
    return {
        ...getServiceOverride$1(),
        ...getServiceOverride$2(),
        [( IConfigurationService.toString())]: new SyncDescriptor(InjectedConfigurationService),
        [( ITextResourceConfigurationService.toString())]: new SyncDescriptor(TextResourceConfigurationService)
    };
}
export { configurationRegistry, getServiceOverride as default, getUserConfiguration, onUserConfigurationChange, updateUserConfiguration };
