// vscode/dist/contextkeys.js
import { registerColor, editorBackground, transparent, contrastBorder, treeIndentGuidesStroke, focusBorder, editorWidgetForeground, editorWidgetBackground, editorForeground, inputBorder, listActiveSelectionBackground, darken, listActiveSelectionForeground, editorInfoForeground, activeContrastBorder, errorForeground, editorWarningForeground, toolbarHoverBackground, widgetBorder, textLinkForeground, lighten, editorErrorForeground } from 'monaco-editor/esm/vs/platform/theme/common/colorRegistry.js';
import { localize } from 'monaco-editor/esm/vs/nls.js';
import { Color } from 'monaco-editor/esm/vs/base/common/color.js';
import { _ as __decorate, a as __param, n as IFileService } from './services2.js';
import { RawContextKey, IContextKeyService } from 'monaco-editor/esm/vs/platform/contextkey/common/contextkey.js';
import { ILanguageService } from 'monaco-editor/esm/vs/editor/common/languages/language.js';
import { IModelService } from 'monaco-editor/esm/vs/editor/common/services/model.js';
import { DisposableStore } from 'monaco-editor/esm/vs/base/common/lifecycle.js';
import { isEqual, basename, dirname, extname } from 'monaco-editor/esm/vs/base/common/resources.js';
import { Schemas } from 'monaco-editor/esm/vs/base/common/network.js';
function isErrorWithActions(obj) {
    const candidate = obj;
    return candidate instanceof Error && Array.isArray(candidate.actions);
}
function createErrorWithActions(messageOrError, actions) {
    let error;
    if (typeof messageOrError === 'string') {
        error = ( new Error(messageOrError));
    }
    else {
        error = messageOrError;
    }
    error.actions = actions;
    return error;
}
const TAB_ACTIVE_BACKGROUND = registerColor('tab.activeBackground', {
    dark: editorBackground,
    light: editorBackground,
    hcDark: editorBackground,
    hcLight: editorBackground
}, ( localize(
    'tabActiveBackground',
    "Active tab background color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedActiveBackground', {
    dark: TAB_ACTIVE_BACKGROUND,
    light: TAB_ACTIVE_BACKGROUND,
    hcDark: TAB_ACTIVE_BACKGROUND,
    hcLight: TAB_ACTIVE_BACKGROUND,
}, ( localize(
    'tabUnfocusedActiveBackground',
    "Active tab background color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_INACTIVE_BACKGROUND = registerColor('tab.inactiveBackground', {
    dark: '#2D2D2D',
    light: '#ECECEC',
    hcDark: null,
    hcLight: null,
}, ( localize(
    'tabInactiveBackground',
    "Inactive tab background color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedInactiveBackground', {
    dark: TAB_INACTIVE_BACKGROUND,
    light: TAB_INACTIVE_BACKGROUND,
    hcDark: TAB_INACTIVE_BACKGROUND,
    hcLight: TAB_INACTIVE_BACKGROUND
}, ( localize(
    'tabUnfocusedInactiveBackground',
    "Inactive tab background color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_FOREGROUND = registerColor('tab.activeForeground', {
    dark: Color.white,
    light: '#333333',
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    'tabActiveForeground',
    "Active tab foreground color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_INACTIVE_FOREGROUND = registerColor('tab.inactiveForeground', {
    dark: ( transparent(TAB_ACTIVE_FOREGROUND, 0.5)),
    light: ( transparent(TAB_ACTIVE_FOREGROUND, 0.7)),
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    'tabInactiveForeground',
    "Inactive tab foreground color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedActiveForeground', {
    dark: ( transparent(TAB_ACTIVE_FOREGROUND, 0.5)),
    light: ( transparent(TAB_ACTIVE_FOREGROUND, 0.7)),
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    'tabUnfocusedActiveForeground',
    "Active tab foreground color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedInactiveForeground', {
    dark: ( transparent(TAB_INACTIVE_FOREGROUND, 0.5)),
    light: ( transparent(TAB_INACTIVE_FOREGROUND, 0.5)),
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    'tabUnfocusedInactiveForeground',
    "Inactive tab foreground color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_HOVER_BACKGROUND = registerColor('tab.hoverBackground', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabHoverBackground',
    "Tab background color when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedHoverBackground', {
    dark: ( transparent(TAB_HOVER_BACKGROUND, 0.5)),
    light: ( transparent(TAB_HOVER_BACKGROUND, 0.7)),
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabUnfocusedHoverBackground',
    "Tab background color in an unfocused group when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_HOVER_FOREGROUND = registerColor('tab.hoverForeground', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null,
}, ( localize(
    'tabHoverForeground',
    "Tab foreground color when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedHoverForeground', {
    dark: ( transparent(TAB_HOVER_FOREGROUND, 0.5)),
    light: ( transparent(TAB_HOVER_FOREGROUND, 0.5)),
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabUnfocusedHoverForeground',
    "Tab foreground color in an unfocused group when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.border', {
    dark: '#252526',
    light: '#F3F3F3',
    hcDark: contrastBorder,
    hcLight: contrastBorder,
}, ( localize(
    'tabBorder',
    "Border to separate tabs from each other. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.lastPinnedBorder', {
    dark: treeIndentGuidesStroke,
    light: treeIndentGuidesStroke,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'lastPinnedTabBorder',
    "Border to separate pinned tabs from other tabs. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_BORDER = registerColor('tab.activeBorder', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabActiveBorder',
    "Border on the bottom of an active tab. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedActiveBorder', {
    dark: ( transparent(TAB_ACTIVE_BORDER, 0.5)),
    light: ( transparent(TAB_ACTIVE_BORDER, 0.7)),
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabActiveUnfocusedBorder',
    "Border on the bottom of an active tab in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_BORDER_TOP = registerColor('tab.activeBorderTop', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    'tabActiveBorderTop',
    "Border to the top of an active tab. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedActiveBorderTop', {
    dark: ( transparent(TAB_ACTIVE_BORDER_TOP, 0.5)),
    light: ( transparent(TAB_ACTIVE_BORDER_TOP, 0.7)),
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    'tabActiveUnfocusedBorderTop',
    "Border to the top of an active tab in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_HOVER_BORDER = registerColor('tab.hoverBorder', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabHoverBorder',
    "Border to highlight tabs when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedHoverBorder', {
    dark: ( transparent(TAB_HOVER_BORDER, 0.5)),
    light: ( transparent(TAB_HOVER_BORDER, 0.7)),
    hcDark: null,
    hcLight: contrastBorder
}, ( localize(
    'tabUnfocusedHoverBorder',
    "Border to highlight tabs in an unfocused group when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_MODIFIED_BORDER = registerColor('tab.activeModifiedBorder', {
    dark: '#3399CC',
    light: '#33AAEE',
    hcDark: null,
    hcLight: contrastBorder
}, ( localize(
    'tabActiveModifiedBorder',
    "Border on the top of modified active tabs in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_INACTIVE_MODIFIED_BORDER = registerColor('tab.inactiveModifiedBorder', {
    dark: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.5)),
    light: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.5)),
    hcDark: Color.white,
    hcLight: contrastBorder
}, ( localize(
    'tabInactiveModifiedBorder',
    "Border on the top of modified inactive tabs in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedActiveModifiedBorder', {
    dark: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.5)),
    light: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.7)),
    hcDark: Color.white,
    hcLight: contrastBorder
}, ( localize(
    'unfocusedActiveModifiedBorder',
    "Border on the top of modified active tabs in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedInactiveModifiedBorder', {
    dark: ( transparent(TAB_INACTIVE_MODIFIED_BORDER, 0.5)),
    light: ( transparent(TAB_INACTIVE_MODIFIED_BORDER, 0.5)),
    hcDark: Color.white,
    hcLight: contrastBorder
}, ( localize(
    'unfocusedINactiveModifiedBorder',
    "Border on the top of modified inactive tabs in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('editorPane.background', {
    dark: editorBackground,
    light: editorBackground,
    hcDark: editorBackground,
    hcLight: editorBackground
}, ( localize(
    'editorPaneBackground',
    "Background color of the editor pane visible on the left and right side of the centered editor layout."
)));
registerColor('editorGroup.emptyBackground', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'editorGroupEmptyBackground',
    "Background color of an empty editor group. Editor groups are the containers of editors."
)));
registerColor('editorGroup.focusedEmptyBorder', {
    dark: null,
    light: null,
    hcDark: focusBorder,
    hcLight: focusBorder
}, ( localize(
    'editorGroupFocusedEmptyBorder',
    "Border color of an empty editor group that is focused. Editor groups are the containers of editors."
)));
registerColor('editorGroupHeader.tabsBackground', {
    dark: '#252526',
    light: '#F3F3F3',
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabsContainerBackground',
    "Background color of the editor group title header when tabs are enabled. Editor groups are the containers of editors."
)));
registerColor('editorGroupHeader.tabsBorder', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'tabsContainerBorder',
    "Border color of the editor group title header when tabs are enabled. Editor groups are the containers of editors."
)));
registerColor('editorGroupHeader.noTabsBackground', {
    dark: editorBackground,
    light: editorBackground,
    hcDark: editorBackground,
    hcLight: editorBackground
}, ( localize(
    'editorGroupHeaderBackground',
    "Background color of the editor group title header when tabs are disabled (`\"workbench.editor.showTabs\": false`). Editor groups are the containers of editors."
)));
registerColor('editorGroupHeader.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'editorTitleContainerBorder',
    "Border color of the editor group title header. Editor groups are the containers of editors."
)));
const EDITOR_GROUP_BORDER = registerColor('editorGroup.border', {
    dark: '#444444',
    light: '#E7E7E7',
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'editorGroupBorder',
    "Color to separate multiple editor groups from each other. Editor groups are the containers of editors."
)));
const EDITOR_DRAG_AND_DROP_BACKGROUND = registerColor('editorGroup.dropBackground', {
    dark: ( ( Color.fromHex('#53595D')).transparent(0.5)),
    light: ( ( Color.fromHex('#2677CB')).transparent(0.18)),
    hcDark: null,
    hcLight: ( ( Color.fromHex('#0F4A85')).transparent(0.50))
}, ( localize(
    'editorDragAndDropBackground',
    "Background color when dragging editors around. The color should have transparency so that the editor contents can still shine through."
)));
registerColor('editorGroup.dropIntoPromptForeground', {
    dark: editorWidgetForeground,
    light: editorWidgetForeground,
    hcDark: editorWidgetForeground,
    hcLight: editorWidgetForeground
}, ( localize(
    'editorDropIntoPromptForeground',
    "Foreground color of text shown over editors when dragging files. This text informs the user that they can hold shift to drop into the editor."
)));
registerColor('editorGroup.dropIntoPromptBackground', {
    dark: editorWidgetBackground,
    light: editorWidgetBackground,
    hcDark: editorWidgetBackground,
    hcLight: editorWidgetBackground
}, ( localize(
    'editorDropIntoPromptBackground',
    "Background color of text shown over editors when dragging files. This text informs the user that they can hold shift to drop into the editor."
)));
registerColor('editorGroup.dropIntoPromptBorder', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'editorDropIntoPromptBorder',
    "Border color of text shown over editors when dragging files. This text informs the user that they can hold shift to drop into the editor."
)));
registerColor('sideBySideEditor.horizontalBorder', {
    dark: EDITOR_GROUP_BORDER,
    light: EDITOR_GROUP_BORDER,
    hcDark: EDITOR_GROUP_BORDER,
    hcLight: EDITOR_GROUP_BORDER
}, ( localize(
    'sideBySideEditor.horizontalBorder',
    "Color to separate two editors from each other when shown side by side in an editor group from top to bottom."
)));
registerColor('sideBySideEditor.verticalBorder', {
    dark: EDITOR_GROUP_BORDER,
    light: EDITOR_GROUP_BORDER,
    hcDark: EDITOR_GROUP_BORDER,
    hcLight: EDITOR_GROUP_BORDER
}, ( localize(
    'sideBySideEditor.verticalBorder',
    "Color to separate two editors from each other when shown side by side in an editor group from left to right."
)));
const PANEL_BACKGROUND = registerColor('panel.background', {
    dark: editorBackground,
    light: editorBackground,
    hcDark: editorBackground,
    hcLight: editorBackground
}, ( localize(
    'panelBackground',
    "Panel background color. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_BORDER = registerColor('panel.border', {
    dark: ( ( Color.fromHex('#808080')).transparent(0.35)),
    light: ( ( Color.fromHex('#808080')).transparent(0.35)),
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'panelBorder',
    "Panel border color to separate the panel from the editor. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_ACTIVE_TITLE_FOREGROUND = registerColor('panelTitle.activeForeground', {
    dark: '#E7E7E7',
    light: '#424242',
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    'panelActiveTitleForeground',
    "Title color for the active panel. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
registerColor('panelTitle.inactiveForeground', {
    dark: ( transparent(PANEL_ACTIVE_TITLE_FOREGROUND, 0.6)),
    light: ( transparent(PANEL_ACTIVE_TITLE_FOREGROUND, 0.75)),
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    'panelInactiveTitleForeground',
    "Title color for the inactive panel. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
registerColor('panelTitle.activeBorder', {
    dark: PANEL_ACTIVE_TITLE_FOREGROUND,
    light: PANEL_ACTIVE_TITLE_FOREGROUND,
    hcDark: contrastBorder,
    hcLight: '#B5200D'
}, ( localize(
    'panelActiveTitleBorder',
    "Border color for the active panel title. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
registerColor('panelInput.border', {
    dark: inputBorder,
    light: ( Color.fromHex('#ddd')),
    hcDark: inputBorder,
    hcLight: inputBorder
}, ( localize('panelInputBorder', "Input box border for inputs in the panel.")));
registerColor('panel.dropBorder', {
    dark: PANEL_ACTIVE_TITLE_FOREGROUND,
    light: PANEL_ACTIVE_TITLE_FOREGROUND,
    hcDark: PANEL_ACTIVE_TITLE_FOREGROUND,
    hcLight: PANEL_ACTIVE_TITLE_FOREGROUND
}, ( localize(
    'panelDragAndDropBorder',
    "Drag and drop feedback color for the panel titles. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
registerColor('panelSection.dropBackground', {
    dark: EDITOR_DRAG_AND_DROP_BACKGROUND,
    light: EDITOR_DRAG_AND_DROP_BACKGROUND,
    hcDark: EDITOR_DRAG_AND_DROP_BACKGROUND,
    hcLight: EDITOR_DRAG_AND_DROP_BACKGROUND
}, ( localize(
    'panelSectionDragAndDropBackground',
    "Drag and drop feedback color for the panel sections. The color should have transparency so that the panel sections can still shine through. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
registerColor('panelSectionHeader.background', {
    dark: ( ( Color.fromHex('#808080')).transparent(0.2)),
    light: ( ( Color.fromHex('#808080')).transparent(0.2)),
    hcDark: null,
    hcLight: null,
}, ( localize(
    'panelSectionHeaderBackground',
    "Panel section header background color. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
registerColor('panelSectionHeader.foreground', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'panelSectionHeaderForeground',
    "Panel section header foreground color. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
registerColor('panelSectionHeader.border', {
    dark: contrastBorder,
    light: contrastBorder,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'panelSectionHeaderBorder',
    "Panel section header border color used when multiple views are stacked vertically in the panel. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
registerColor('panelSection.border', {
    dark: PANEL_BORDER,
    light: PANEL_BORDER,
    hcDark: PANEL_BORDER,
    hcLight: PANEL_BORDER
}, ( localize(
    'panelSectionBorder',
    "Panel section border color used when multiple views are stacked horizontally in the panel. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
registerColor('banner.background', {
    dark: listActiveSelectionBackground,
    light: ( darken(listActiveSelectionBackground, 0.3)),
    hcDark: listActiveSelectionBackground,
    hcLight: listActiveSelectionBackground
}, ( localize(
    'banner.background',
    "Banner background color. The banner is shown under the title bar of the window."
)));
registerColor('banner.foreground', {
    dark: listActiveSelectionForeground,
    light: listActiveSelectionForeground,
    hcDark: listActiveSelectionForeground,
    hcLight: listActiveSelectionForeground
}, ( localize(
    'banner.foreground',
    "Banner foreground color. The banner is shown under the title bar of the window."
)));
registerColor('banner.iconForeground', {
    dark: editorInfoForeground,
    light: editorInfoForeground,
    hcDark: editorInfoForeground,
    hcLight: editorInfoForeground
}, ( localize(
    'banner.iconForeground',
    "Banner icon color. The banner is shown under the title bar of the window."
)));
const STATUS_BAR_FOREGROUND = registerColor('statusBar.foreground', {
    dark: '#FFFFFF',
    light: '#FFFFFF',
    hcDark: '#FFFFFF',
    hcLight: editorForeground
}, ( localize(
    'statusBarForeground',
    "Status bar foreground color when a workspace or folder is opened. The status bar is shown in the bottom of the window."
)));
registerColor('statusBar.noFolderForeground', {
    dark: STATUS_BAR_FOREGROUND,
    light: STATUS_BAR_FOREGROUND,
    hcDark: STATUS_BAR_FOREGROUND,
    hcLight: STATUS_BAR_FOREGROUND
}, ( localize(
    'statusBarNoFolderForeground',
    "Status bar foreground color when no folder is opened. The status bar is shown in the bottom of the window."
)));
registerColor('statusBar.background', {
    dark: '#007ACC',
    light: '#007ACC',
    hcDark: null,
    hcLight: null,
}, ( localize(
    'statusBarBackground',
    "Status bar background color when a workspace or folder is opened. The status bar is shown in the bottom of the window."
)));
registerColor('statusBar.noFolderBackground', {
    dark: '#68217A',
    light: '#68217A',
    hcDark: null,
    hcLight: null,
}, ( localize(
    'statusBarNoFolderBackground',
    "Status bar background color when no folder is opened. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_BORDER = registerColor('statusBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'statusBarBorder',
    "Status bar border color separating to the sidebar and editor. The status bar is shown in the bottom of the window."
)));
registerColor('statusBar.focusBorder', {
    dark: STATUS_BAR_FOREGROUND,
    light: STATUS_BAR_FOREGROUND,
    hcDark: null,
    hcLight: STATUS_BAR_FOREGROUND
}, ( localize(
    'statusBarFocusBorder',
    "Status bar border color when focused on keyboard navigation. The status bar is shown in the bottom of the window."
)));
registerColor('statusBar.noFolderBorder', {
    dark: STATUS_BAR_BORDER,
    light: STATUS_BAR_BORDER,
    hcDark: STATUS_BAR_BORDER,
    hcLight: STATUS_BAR_BORDER
}, ( localize(
    'statusBarNoFolderBorder',
    "Status bar border color separating to the sidebar and editor when no folder is opened. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.activeBackground', {
    dark: ( Color.white.transparent(0.18)),
    light: ( Color.white.transparent(0.18)),
    hcDark: ( Color.white.transparent(0.18)),
    hcLight: ( Color.black.transparent(0.18))
}, ( localize(
    'statusBarItemActiveBackground',
    "Status bar item background color when clicking. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.focusBorder', {
    dark: STATUS_BAR_FOREGROUND,
    light: STATUS_BAR_FOREGROUND,
    hcDark: null,
    hcLight: activeContrastBorder
}, ( localize(
    'statusBarItemFocusBorder',
    "Status bar item border color when focused on keyboard navigation. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.hoverBackground', {
    dark: ( Color.white.transparent(0.12)),
    light: ( Color.white.transparent(0.12)),
    hcDark: ( Color.white.transparent(0.12)),
    hcLight: ( Color.black.transparent(0.12))
}, ( localize(
    'statusBarItemHoverBackground',
    "Status bar item background color when hovering. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.compactHoverBackground', {
    dark: ( Color.white.transparent(0.20)),
    light: ( Color.white.transparent(0.20)),
    hcDark: ( Color.white.transparent(0.20)),
    hcLight: ( Color.black.transparent(0.20))
}, ( localize(
    'statusBarItemCompactHoverBackground',
    "Status bar item background color when hovering an item that contains two hovers. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.prominentForeground', {
    dark: STATUS_BAR_FOREGROUND,
    light: STATUS_BAR_FOREGROUND,
    hcDark: STATUS_BAR_FOREGROUND,
    hcLight: STATUS_BAR_FOREGROUND
}, ( localize(
    'statusBarProminentItemForeground',
    "Status bar prominent items foreground color. Prominent items stand out from other status bar entries to indicate importance. Change mode `Toggle Tab Key Moves Focus` from command palette to see an example. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.prominentBackground', {
    dark: ( Color.black.transparent(0.5)),
    light: ( Color.black.transparent(0.5)),
    hcDark: ( Color.black.transparent(0.5)),
    hcLight: ( Color.black.transparent(0.5)),
}, ( localize(
    'statusBarProminentItemBackground',
    "Status bar prominent items background color. Prominent items stand out from other status bar entries to indicate importance. Change mode `Toggle Tab Key Moves Focus` from command palette to see an example. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.prominentHoverBackground', {
    dark: ( Color.black.transparent(0.3)),
    light: ( Color.black.transparent(0.3)),
    hcDark: ( Color.black.transparent(0.3)),
    hcLight: null
}, ( localize(
    'statusBarProminentItemHoverBackground',
    "Status bar prominent items background color when hovering. Prominent items stand out from other status bar entries to indicate importance. Change mode `Toggle Tab Key Moves Focus` from command palette to see an example. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.errorBackground', {
    dark: ( darken(errorForeground, .4)),
    light: ( darken(errorForeground, .4)),
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    'statusBarErrorItemBackground',
    "Status bar error items background color. Error items stand out from other status bar entries to indicate error conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.errorForeground', {
    dark: Color.white,
    light: Color.white,
    hcDark: Color.white,
    hcLight: Color.white
}, ( localize(
    'statusBarErrorItemForeground',
    "Status bar error items foreground color. Error items stand out from other status bar entries to indicate error conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.warningBackground', {
    dark: ( darken(editorWarningForeground, .4)),
    light: ( darken(editorWarningForeground, .4)),
    hcDark: null,
    hcLight: '#895503'
}, ( localize(
    'statusBarWarningItemBackground',
    "Status bar warning items background color. Warning items stand out from other status bar entries to indicate warning conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.warningForeground', {
    dark: Color.white,
    light: Color.white,
    hcDark: Color.white,
    hcLight: Color.white
}, ( localize(
    'statusBarWarningItemForeground',
    "Status bar warning items foreground color. Warning items stand out from other status bar entries to indicate warning conditions. The status bar is shown in the bottom of the window."
)));
registerColor('activityBar.background', {
    dark: '#333333',
    light: '#2C2C2C',
    hcDark: '#000000',
    hcLight: '#FFFFFF'
}, ( localize(
    'activityBarBackground',
    "Activity bar background color. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_FOREGROUND = registerColor('activityBar.foreground', {
    dark: Color.white,
    light: Color.white,
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    'activityBarForeground',
    "Activity bar item foreground color when it is active. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
registerColor('activityBar.inactiveForeground', {
    dark: ( transparent(ACTIVITY_BAR_FOREGROUND, 0.4)),
    light: ( transparent(ACTIVITY_BAR_FOREGROUND, 0.4)),
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    'activityBarInActiveForeground',
    "Activity bar item foreground color when it is inactive. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
registerColor('activityBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'activityBarBorder',
    "Activity bar border color separating to the side bar. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
registerColor('activityBar.activeBorder', {
    dark: ACTIVITY_BAR_FOREGROUND,
    light: ACTIVITY_BAR_FOREGROUND,
    hcDark: null,
    hcLight: contrastBorder
}, ( localize(
    'activityBarActiveBorder',
    "Activity bar border color for the active item. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
registerColor('activityBar.activeFocusBorder', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    'activityBarActiveFocusBorder',
    "Activity bar focus border color for the active item. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
registerColor('activityBar.activeBackground', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'activityBarActiveBackground',
    "Activity bar background color for the active item. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
registerColor('activityBar.dropBorder', {
    dark: ACTIVITY_BAR_FOREGROUND,
    light: ACTIVITY_BAR_FOREGROUND,
    hcDark: null,
    hcLight: null,
}, ( localize(
    'activityBarDragAndDropBorder',
    "Drag and drop feedback color for the activity bar items. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_BADGE_BACKGROUND = registerColor('activityBarBadge.background', {
    dark: '#007ACC',
    light: '#007ACC',
    hcDark: '#000000',
    hcLight: '#0F4A85'
}, ( localize(
    'activityBarBadgeBackground',
    "Activity notification badge background color. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_BADGE_FOREGROUND = registerColor('activityBarBadge.foreground', {
    dark: Color.white,
    light: Color.white,
    hcDark: Color.white,
    hcLight: Color.white
}, ( localize(
    'activityBarBadgeForeground',
    "Activity notification badge foreground color. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
registerColor('profileBadge.background', {
    dark: '#4D4D4D',
    light: '#C4C4C4',
    hcDark: Color.white,
    hcLight: Color.black
}, ( localize(
    'profileBadgeBackground',
    "Profile badge background color. The profile badge shows on top of the settings gear icon in the activity bar."
)));
registerColor('profileBadge.foreground', {
    dark: Color.white,
    light: '#333333',
    hcDark: Color.black,
    hcLight: Color.white
}, ( localize(
    'profileBadgeForeground',
    "Profile badge foreground color. The profile badge shows on top of the settings gear icon in the activity bar."
)));
registerColor('statusBarItem.remoteBackground', {
    dark: ACTIVITY_BAR_BADGE_BACKGROUND,
    light: ACTIVITY_BAR_BADGE_BACKGROUND,
    hcDark: ACTIVITY_BAR_BADGE_BACKGROUND,
    hcLight: ACTIVITY_BAR_BADGE_BACKGROUND
}, ( localize(
    'statusBarItemHostBackground',
    "Background color for the remote indicator on the status bar."
)));
registerColor('statusBarItem.remoteForeground', {
    dark: ACTIVITY_BAR_BADGE_FOREGROUND,
    light: ACTIVITY_BAR_BADGE_FOREGROUND,
    hcDark: ACTIVITY_BAR_BADGE_FOREGROUND,
    hcLight: ACTIVITY_BAR_BADGE_FOREGROUND
}, ( localize(
    'statusBarItemHostForeground',
    "Foreground color for the remote indicator on the status bar."
)));
registerColor('extensionBadge.remoteBackground', {
    dark: ACTIVITY_BAR_BADGE_BACKGROUND,
    light: ACTIVITY_BAR_BADGE_BACKGROUND,
    hcDark: ACTIVITY_BAR_BADGE_BACKGROUND,
    hcLight: ACTIVITY_BAR_BADGE_BACKGROUND
}, ( localize(
    'extensionBadge.remoteBackground',
    "Background color for the remote badge in the extensions view."
)));
registerColor('extensionBadge.remoteForeground', {
    dark: ACTIVITY_BAR_BADGE_FOREGROUND,
    light: ACTIVITY_BAR_BADGE_FOREGROUND,
    hcDark: ACTIVITY_BAR_BADGE_FOREGROUND,
    hcLight: ACTIVITY_BAR_BADGE_FOREGROUND
}, ( localize(
    'extensionBadge.remoteForeground',
    "Foreground color for the remote badge in the extensions view."
)));
const SIDE_BAR_BACKGROUND = registerColor('sideBar.background', {
    dark: '#252526',
    light: '#F3F3F3',
    hcDark: '#000000',
    hcLight: '#FFFFFF'
}, ( localize(
    'sideBarBackground',
    "Side bar background color. The side bar is the container for views like explorer and search."
)));
const SIDE_BAR_FOREGROUND = registerColor('sideBar.foreground', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'sideBarForeground',
    "Side bar foreground color. The side bar is the container for views like explorer and search."
)));
registerColor('sideBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'sideBarBorder',
    "Side bar border color on the side separating to the editor. The side bar is the container for views like explorer and search."
)));
registerColor('sideBarTitle.foreground', {
    dark: SIDE_BAR_FOREGROUND,
    light: SIDE_BAR_FOREGROUND,
    hcDark: SIDE_BAR_FOREGROUND,
    hcLight: SIDE_BAR_FOREGROUND
}, ( localize(
    'sideBarTitleForeground',
    "Side bar title foreground color. The side bar is the container for views like explorer and search."
)));
registerColor('sideBar.dropBackground', {
    dark: EDITOR_DRAG_AND_DROP_BACKGROUND,
    light: EDITOR_DRAG_AND_DROP_BACKGROUND,
    hcDark: EDITOR_DRAG_AND_DROP_BACKGROUND,
    hcLight: EDITOR_DRAG_AND_DROP_BACKGROUND
}, ( localize(
    'sideBarDragAndDropBackground',
    "Drag and drop feedback color for the side bar sections. The color should have transparency so that the side bar sections can still shine through. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
registerColor('sideBarSectionHeader.background', {
    dark: ( ( Color.fromHex('#808080')).transparent(0.2)),
    light: ( ( Color.fromHex('#808080')).transparent(0.2)),
    hcDark: null,
    hcLight: null
}, ( localize(
    'sideBarSectionHeaderBackground',
    "Side bar section header background color. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
registerColor('sideBarSectionHeader.foreground', {
    dark: SIDE_BAR_FOREGROUND,
    light: SIDE_BAR_FOREGROUND,
    hcDark: SIDE_BAR_FOREGROUND,
    hcLight: SIDE_BAR_FOREGROUND
}, ( localize(
    'sideBarSectionHeaderForeground',
    "Side bar section header foreground color. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
registerColor('sideBarSectionHeader.border', {
    dark: contrastBorder,
    light: contrastBorder,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'sideBarSectionHeaderBorder',
    "Side bar section header border color. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
const TITLE_BAR_ACTIVE_FOREGROUND = registerColor('titleBar.activeForeground', {
    dark: '#CCCCCC',
    light: '#333333',
    hcDark: '#FFFFFF',
    hcLight: '#292929'
}, ( localize(
    'titleBarActiveForeground',
    "Title bar foreground when the window is active."
)));
registerColor('titleBar.inactiveForeground', {
    dark: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, 0.6)),
    light: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, 0.6)),
    hcDark: null,
    hcLight: '#292929'
}, ( localize(
    'titleBarInactiveForeground',
    "Title bar foreground when the window is inactive."
)));
const TITLE_BAR_ACTIVE_BACKGROUND = registerColor('titleBar.activeBackground', {
    dark: '#3C3C3C',
    light: '#DDDDDD',
    hcDark: '#000000',
    hcLight: '#FFFFFF'
}, ( localize(
    'titleBarActiveBackground',
    "Title bar background when the window is active."
)));
registerColor('titleBar.inactiveBackground', {
    dark: ( transparent(TITLE_BAR_ACTIVE_BACKGROUND, 0.6)),
    light: ( transparent(TITLE_BAR_ACTIVE_BACKGROUND, 0.6)),
    hcDark: null,
    hcLight: null,
}, ( localize(
    'titleBarInactiveBackground',
    "Title bar background when the window is inactive."
)));
registerColor('titleBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize('titleBarBorder', "Title bar border color.")));
registerColor('menubar.selectionForeground', {
    dark: TITLE_BAR_ACTIVE_FOREGROUND,
    light: TITLE_BAR_ACTIVE_FOREGROUND,
    hcDark: TITLE_BAR_ACTIVE_FOREGROUND,
    hcLight: TITLE_BAR_ACTIVE_FOREGROUND,
}, ( localize(
    'menubarSelectionForeground',
    "Foreground color of the selected menu item in the menubar."
)));
registerColor('menubar.selectionBackground', {
    dark: toolbarHoverBackground,
    light: toolbarHoverBackground,
    hcDark: null,
    hcLight: null,
}, ( localize(
    'menubarSelectionBackground',
    "Background color of the selected menu item in the menubar."
)));
registerColor('menubar.selectionBorder', {
    dark: null,
    light: null,
    hcDark: activeContrastBorder,
    hcLight: activeContrastBorder,
}, ( localize(
    'menubarSelectionBorder',
    "Border color of the selected menu item in the menubar."
)));
const NOTIFICATIONS_CENTER_BORDER = registerColor('notificationCenter.border', {
    dark: widgetBorder,
    light: widgetBorder,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'notificationCenterBorder',
    "Notifications center border color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_TOAST_BORDER = registerColor('notificationToast.border', {
    dark: widgetBorder,
    light: widgetBorder,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'notificationToastBorder',
    "Notification toast border color. Notifications slide in from the bottom right of the window."
)));
registerColor('notifications.foreground', {
    dark: editorWidgetForeground,
    light: editorWidgetForeground,
    hcDark: editorWidgetForeground,
    hcLight: editorWidgetForeground
}, ( localize(
    'notificationsForeground',
    "Notifications foreground color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_BACKGROUND = registerColor('notifications.background', {
    dark: editorWidgetBackground,
    light: editorWidgetBackground,
    hcDark: editorWidgetBackground,
    hcLight: editorWidgetBackground
}, ( localize(
    'notificationsBackground',
    "Notifications background color. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationLink.foreground', {
    dark: textLinkForeground,
    light: textLinkForeground,
    hcDark: textLinkForeground,
    hcLight: textLinkForeground
}, ( localize(
    'notificationsLink',
    "Notification links foreground color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_CENTER_HEADER_FOREGROUND = registerColor('notificationCenterHeader.foreground', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: null
}, ( localize(
    'notificationCenterHeaderForeground',
    "Notifications center header foreground color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_CENTER_HEADER_BACKGROUND = registerColor('notificationCenterHeader.background', {
    dark: ( lighten(NOTIFICATIONS_BACKGROUND, 0.3)),
    light: ( darken(NOTIFICATIONS_BACKGROUND, 0.05)),
    hcDark: NOTIFICATIONS_BACKGROUND,
    hcLight: NOTIFICATIONS_BACKGROUND
}, ( localize(
    'notificationCenterHeaderBackground',
    "Notifications center header background color. Notifications slide in from the bottom right of the window."
)));
registerColor('notifications.border', {
    dark: NOTIFICATIONS_CENTER_HEADER_BACKGROUND,
    light: NOTIFICATIONS_CENTER_HEADER_BACKGROUND,
    hcDark: NOTIFICATIONS_CENTER_HEADER_BACKGROUND,
    hcLight: NOTIFICATIONS_CENTER_HEADER_BACKGROUND
}, ( localize(
    'notificationsBorder',
    "Notifications border color separating from other notifications in the notifications center. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationsErrorIcon.foreground', {
    dark: editorErrorForeground,
    light: editorErrorForeground,
    hcDark: editorErrorForeground,
    hcLight: editorErrorForeground
}, ( localize(
    'notificationsErrorIconForeground',
    "The color used for the icon of error notifications. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationsWarningIcon.foreground', {
    dark: editorWarningForeground,
    light: editorWarningForeground,
    hcDark: editorWarningForeground,
    hcLight: editorWarningForeground
}, ( localize(
    'notificationsWarningIconForeground',
    "The color used for the icon of warning notifications. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationsInfoIcon.foreground', {
    dark: editorInfoForeground,
    light: editorInfoForeground,
    hcDark: editorInfoForeground,
    hcLight: editorInfoForeground
}, ( localize(
    'notificationsInfoIconForeground',
    "The color used for the icon of info notifications. Notifications slide in from the bottom right of the window."
)));
registerColor('window.activeBorder', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'windowActiveBorder',
    "The color used for the border of the window when it is active. Only supported in the macOS and Linux desktop client when using the custom title bar."
)));
registerColor('window.inactiveBorder', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    'windowInactiveBorder',
    "The color used for the border of the window when it is inactive. Only supported in the macOS and Linux desktop client when using the custom title bar."
)));
const NotificationFocusedContext = ( new RawContextKey('notificationFocus', true, ( localize('notificationFocus', "Whether a notification has keyboard focus"))));
const NotificationsCenterVisibleContext = ( new RawContextKey('notificationCenterVisible', false, ( localize('notificationCenterVisible', "Whether the notifications center is visible"))));
const NotificationsToastsVisibleContext = ( new RawContextKey('notificationToastsVisible', false, ( localize('notificationToastsVisible', "Whether a notification toast is visible"))));
const PanelFocusContext = ( new RawContextKey('panelFocus', false, ( localize('panelFocus', "Whether the panel has keyboard focus"))));
let ResourceContextKey = class ResourceContextKey {
    constructor(_contextKeyService, _fileService, _languageService, _modelService) {
        this._contextKeyService = _contextKeyService;
        this._fileService = _fileService;
        this._languageService = _languageService;
        this._modelService = _modelService;
        this._disposables = ( new DisposableStore());
        this._schemeKey = ResourceContextKey.Scheme.bindTo(this._contextKeyService);
        this._filenameKey = ResourceContextKey.Filename.bindTo(this._contextKeyService);
        this._dirnameKey = ResourceContextKey.Dirname.bindTo(this._contextKeyService);
        this._pathKey = ResourceContextKey.Path.bindTo(this._contextKeyService);
        this._langIdKey = ResourceContextKey.LangId.bindTo(this._contextKeyService);
        this._resourceKey = ResourceContextKey.Resource.bindTo(this._contextKeyService);
        this._extensionKey = ResourceContextKey.Extension.bindTo(this._contextKeyService);
        this._hasResource = ResourceContextKey.HasResource.bindTo(this._contextKeyService);
        this._isFileSystemResource = ResourceContextKey.IsFileSystemResource.bindTo(this._contextKeyService);
        this._disposables.add(_fileService.onDidChangeFileSystemProviderRegistrations(() => {
            const resource = this.get();
            this._isFileSystemResource.set(Boolean(resource && _fileService.hasProvider(resource)));
        }));
        this._disposables.add(_modelService.onModelAdded(model => {
            if (isEqual(model.uri, this.get())) {
                this._setLangId();
            }
        }));
        this._disposables.add(_modelService.onModelLanguageChanged(e => {
            if (isEqual(e.model.uri, this.get())) {
                this._setLangId();
            }
        }));
    }
    dispose() {
        this._disposables.dispose();
    }
    _setLangId() {
        const value = this.get();
        if (!value) {
            this._langIdKey.set(null);
            return;
        }
        const langId = this._modelService.getModel(value)?.getLanguageId() ?? this._languageService.guessLanguageIdByFilepathOrFirstLine(value);
        this._langIdKey.set(langId);
    }
    set(value) {
        value = value ?? undefined;
        if (isEqual(this._value, value)) {
            return;
        }
        this._value = value;
        this._contextKeyService.bufferChangeEvents(() => {
            this._resourceKey.set(value ? ( value.toString()) : null);
            this._schemeKey.set(value ? value.scheme : null);
            this._filenameKey.set(value ? basename(value) : null);
            this._dirnameKey.set(value ? this.uriToPath(dirname(value)) : null);
            this._pathKey.set(value ? this.uriToPath(value) : null);
            this._setLangId();
            this._extensionKey.set(value ? extname(value) : null);
            this._hasResource.set(Boolean(value));
            this._isFileSystemResource.set(value ? this._fileService.hasProvider(value) : false);
        });
    }
    uriToPath(uri) {
        if (uri.scheme === Schemas.file) {
            return uri.fsPath;
        }
        return uri.path;
    }
    reset() {
        this._value = undefined;
        this._contextKeyService.bufferChangeEvents(() => {
            this._resourceKey.reset();
            this._schemeKey.reset();
            this._filenameKey.reset();
            this._dirnameKey.reset();
            this._pathKey.reset();
            this._langIdKey.reset();
            this._extensionKey.reset();
            this._hasResource.reset();
            this._isFileSystemResource.reset();
        });
    }
    get() {
        return this._value;
    }
};
ResourceContextKey.Scheme = ( new RawContextKey('resourceScheme', undefined, { type: 'string', description: ( localize('resourceScheme', "The scheme of the resource")) }));
ResourceContextKey.Filename = ( new RawContextKey(
    'resourceFilename',
    undefined,
    { type: 'string', description: ( localize('resourceFilename', "The file name of the resource")) }
));
ResourceContextKey.Dirname = ( new RawContextKey(
    'resourceDirname',
    undefined,
    { type: 'string', description: ( localize('resourceDirname', "The folder name the resource is contained in")) }
));
ResourceContextKey.Path = ( new RawContextKey('resourcePath', undefined, { type: 'string', description: ( localize('resourcePath', "The full path of the resource")) }));
ResourceContextKey.LangId = ( new RawContextKey('resourceLangId', undefined, { type: 'string', description: ( localize('resourceLangId', "The language identifier of the resource")) }));
ResourceContextKey.Resource = ( new RawContextKey('resource', undefined, { type: 'URI', description: ( localize('resource', "The full value of the resource including scheme and path")) }));
ResourceContextKey.Extension = ( new RawContextKey(
    'resourceExtname',
    undefined,
    { type: 'string', description: ( localize('resourceExtname', "The extension name of the resource")) }
));
ResourceContextKey.HasResource = ( new RawContextKey('resourceSet', undefined, { type: 'boolean', description: ( localize('resourceSet', "Whether a resource is present or not")) }));
ResourceContextKey.IsFileSystemResource = ( new RawContextKey(
    'isFileSystemResource',
    undefined,
    { type: 'boolean', description: ( localize(
        'isFileSystemResource',
        "Whether the resource is backed by a file system provider"
    )) }
));
ResourceContextKey = ( __decorate([
    ( __param(0, IContextKeyService)),
    ( __param(1, IFileService)),
    ( __param(2, ILanguageService)),
    ( __param(3, IModelService))
], ResourceContextKey));
export { NotificationsToastsVisibleContext as N, PanelFocusContext as P, SIDE_BAR_BACKGROUND as S, NOTIFICATIONS_BACKGROUND as a, NOTIFICATIONS_TOAST_BORDER as b, NotificationsCenterVisibleContext as c, NOTIFICATIONS_CENTER_BORDER as d, NOTIFICATIONS_CENTER_HEADER_FOREGROUND as e, NOTIFICATIONS_CENTER_HEADER_BACKGROUND as f, NotificationFocusedContext as g, PANEL_BACKGROUND as h, isErrorWithActions as i, createErrorWithActions as j };
