// monaco-languageclient/lib/console-window.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) 2018-2022 TypeFox GmbH (http://www.typefox.io). All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
import { Severity } from 'vscode/services';
import * as vscode from 'vscode';
export class ConsoleWindow {
    channels = new Map();
    showMessage(severity, message, ..._actions) {
        if (severity === Severity.Error) {
            console.error(message);
        }
        if (severity === Severity.Warning) {
            console.warn(message);
        }
        if (severity === Severity.Info) {
            console.info(message);
        }
        if (severity === Severity.Ignore) {
            console.debug(message);
        }
        return Promise.resolve(undefined);
    }
    createOutputChannel(name, _options) {
        const existing = this.channels.get(name);
        if (existing) {
            return existing;
        }
        const channel = {
            name: 'default',
            logLevel: vscode.LogLevel.Info,
            onDidChangeLogLevel: new vscode.EventEmitter().event,
            trace(_message, ..._args) {
                // no-op
            },
            debug(_message, ..._args) {
                // no-op
            },
            info(_message, ..._args) {
                // no-op
            },
            warn(_message, ..._args) {
                // no-op
            },
            error(_message, ..._args) {
                // no-op
            },
            append(value) {
                console.log(name + ': ' + value);
            },
            appendLine(line) {
                console.log(name + ': ' + line);
            },
            show() {
                // no-op
            },
            dispose() {
                // no-op
            },
            replace: function (_value) {
                // no-op
            },
            clear: function () {
                // no-op
            },
            hide: function () {
                // no-op
            }
        };
        this.channels.set(name, channel);
        return channel;
    }
}
